File: src/Classes/AddressBookDropView.m
#import "AddressBookDropView.h"
#import "XJPreferences.h"

@implementation AddressBookDropView

- (void)awakeFromNib
{
    [self registerForDraggedTypes: kDragTypesArray];
	dragState = YES;
}

- (void)setAcceptsDrags: (BOOL)accept
{
	if(accept) {
		if(!dragState)
			[self registerForDraggedTypes: kDragTypesArray];
	}
	else {
		if(dragState)
			[self unregisterDraggedTypes];
	}
	dragState = accept;
}
- (BOOL)acceptsDrags { return dragState; }

- (NSDragOperation)draggingEntered: (id <NSDraggingInfo>)sender
{
    return NSDragOperationCopy;
}

- (BOOL)prepareForDragOperation: (id <NSDraggingInfo>)sender
{
    return YES;
}

- (BOOL)performDragOperation: (id <NSDraggingInfo>)sender
{
    NSPasteboard *pb = [sender draggingPasteboard];
    NSArray *uidArrays;

    NSString *pbString = [pb stringForType: @"ABPeopleUIDsPboardType"];

    uidArrays = (NSArray *)[pbString propertyList];

    [[NSNotificationCenter defaultCenter] postNotificationName: XJAddressCardDroppedNotification
                                                        object: uidArrays];
    
    return YES;
}

- (void)concludeDragOperation: (id <NSDraggingInfo>)sender
{

}
@end

File: src/Classes/AddressBookDropView.swift
//
//  AddressBookDropView.swift
//  Xjournal
//
//  Created by C.W. Betts on 3/26/15.
//
//

import Cocoa

private let kABDragType = "ABPeopleUIDsPboardType"
private let kDragTypes = [kABDragType]

class AddressBookDropView: NSImageView {

    var acceptsDrags: Bool = false {
        willSet {
            if newValue {
                if !acceptsDrags {
                    register(forDraggedTypes: kDragTypes)
                }
            } else {
                if acceptsDrags {
                    unregisterDraggedTypes()
                }
            }
        }
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        register(forDraggedTypes: kDragTypes)
        acceptsDrags = true
    }
    
    override func prepareForDragOperation(_ sender: NSDraggingInfo) -> Bool {
        return true
    }
    
    override func draggingEntered(_ sender: NSDraggingInfo) -> NSDragOperation {
        return .copy
    }
    
    override func performDragOperation(_ sender: NSDraggingInfo) -> Bool {
        let pb = sender.draggingPasteboard()
        if let uidArrays = pb.propertyList(forType: kABDragType) as? NSArray {
            NotificationCenter.default.post(name: .XJAddressCardDropped, object: uidArrays)
        }
        
        return true
    }
    
    override func concludeDragOperation(_ sender: NSDraggingInfo?) {
        
    }
}
